File: memory-gateway/app/providers/memory_ai.py
"""Memory AI adapter (pluggable cheap model for extraction and compression)."""

from __future__ import annotations

import json
import logging
import re
from typing import Any

import httpx
from pydantic import ValidationError

from app.config import Settings, get_settings
from app.models.memory import MemoryAIOutput, ToolSummaryOutput

logger = logging.getLogger(__name__)

# Pattern to extract JSON payload even if wrapped in markdown code blocks
_JSON_BLOCK_RE = re.compile(
    r"```(?:json)?\s*(?P<json>\{.*\}|\[.*\])\s*```",
    re.DOTALL,
)

EXTRACTION_SYSTEM_PROMPT = """You are an AI Memory Compressor. Your role is ONLY to extract and compress memory items from conversation messages.
You NEVER answer the user directly.

Output strictly valid JSON conforming to this schema:
{
  "summary": "Brief 1-2 sentence overview of what changed or happened",
  "facts": ["Stable facts about project/user/environment"],
  "decisions": ["Confirmed choices and decisions"],
  "constraints": ["Rules, restrictions, and negative requirements"],
  "preferences": ["User preferences and stylistic choices"],
  "goals": ["Active objectives and desired outcomes"],
  "architecture": ["System design and technical stack details"],
  "important_events": ["Key milestones, releases, or deployments"],
  "active_tasks": ["Pending or in-progress todos and tasks"],
  "obsolete_items": ["Previous facts/decisions that are now obsolete"],
  "contradictions": ["Previous items contradicted or superseded by new input"],
  "confidence": 0.95
}

Rules:
1. Extract only high-value information.
2. User statements and explicit decisions have highest authority.
3. Assistant speculation must NOT be recorded as confirmed facts/decisions.
4. Output ONLY valid JSON. No commentary outside the JSON.
"""

TOOL_COMPRESSION_SYSTEM_PROMPT = """You are a Tool Output Compressor. Your role is to summarize tool/command execution output.
Keep crucial exit statuses, error codes, file paths, IDs, and core outputs.
Remove repetitive logs, progress bars, and boilerplate.

Output strictly valid JSON:
{
  "summary": "Concise 1-3 sentence summary of the execution outcome",
  "key_points": ["Specific result item, path, or error"],
  "status": "success | error | warning"
}
"""


def extract_json_text(text: str) -> str:
    """Extract raw JSON substring from model output, handling code fences."""
    trimmed = text.strip()
    match = _JSON_BLOCK_RE.search(trimmed)
    if match:
        return match.group("json").strip()
    first_brace = trimmed.find("{")
    last_brace = trimmed.rfind("}")
    if first_brace != -1 and last_brace != -1 and last_brace > first_brace:
        return trimmed[first_brace : last_brace + 1]
    return trimmed


class MemoryAIAdapter:
    """
    Adapter for invoking a cheap Memory AI model.

    Designed with failure isolation: failures never raise out to break the hot path;
    instead they log and return None / fallback.
    """

    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        model: str,
        timeout: float = 30.0,
        retry_once: bool = True,
        client: httpx.AsyncClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._model = model
        self._timeout = timeout
        self._retry_once = retry_once
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(
            base_url=self._base_url,
            timeout=httpx.Timeout(timeout, connect=10.0),
        )

    def _auth_headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"
        return headers

    async def _post_chat(self, messages: list[dict[str, str]]) -> str:
        url = "/chat/completions"
        payload: dict[str, Any] = {
            "model": self._model,
            "messages": messages,
            "temperature": 0.0,
        }
        response = await self._client.post(
            url,
            json=payload,
            headers=self._auth_headers(),
        )
        response.raise_for_status()
        data = response.json()
        choices = data.get("choices") or []
        if not choices:
            raise ValueError("No completion choices returned by Memory AI")
        content = choices[0].get("message", {}).get("content", "")
        return str(content)

    async def extract_memory(
        self,
        messages_text: str,
        current_memory_summary: str | None = None,
    ) -> MemoryAIOutput | None:
        """
        Call Memory AI to extract structured items.

        Returns MemoryAIOutput on success, or None on failure (preserving prior state).
        """
        user_prompt = f"Delta Messages to analyze:\n{messages_text}"
        if current_memory_summary:
            user_prompt = f"Current Memory State:\n{current_memory_summary}\n\n{user_prompt}"

        conversation: list[dict[str, str]] = [
            {"role": "system", "content": EXTRACTION_SYSTEM_PROMPT},
            {"role": "user", "content": user_prompt},
        ]

        # First attempt
        raw_response: str | None = None
        try:
            raw_response = await self._post_chat(conversation)
            return self._parse_and_validate_memory(raw_response)
        except (ValidationError, json.JSONDecodeError, ValueError) as parse_err:
            if not self._retry_once:
                logger.warning("Memory AI output parsing failed: %s; keeping prior memory", parse_err)
                return None

            logger.info("Memory AI output parse failed (%s); retrying once with schema reminder", parse_err)
            # Retry once
            try:
                retry_conv = list(conversation)
                if raw_response:
                    retry_conv.append({"role": "assistant", "content": raw_response})
                retry_conv.append({
                    "role": "user",
                    "content": (
                        f"Previous output was not valid JSON matching the schema: {parse_err}. "
                        "Please output ONLY the valid JSON object without any additional text."
                    ),
                })
                retry_raw = await self._post_chat(retry_conv)
                return self._parse_and_validate_memory(retry_raw)
            except Exception as retry_err:
                logger.warning(
                    "Memory AI retry also failed: %s; keeping prior memory unchanged",
                    retry_err,
                )
                return None
        except Exception as exc:
            logger.warning("Memory AI request failed: %s; keeping prior memory unchanged", exc)
            return None

    def _parse_and_validate_memory(self, raw_text: str) -> MemoryAIOutput:
        clean = extract_json_text(raw_text)
        return MemoryAIOutput.model_validate_json(clean)

    async def compress_tool_output(
        self,
        tool_name: str,
        output_text: str,
        max_chars_threshold: int = 300,
    ) -> str:
        """
        Compress tool output into a concise summary.

        If output is already short or Memory AI fails, returns safe fallback.
        """
        if len(output_text) <= max_chars_threshold:
            return output_text

        prompt = f"Tool Name: {tool_name}\nRaw Output:\n{output_text}"
        conv = [
            {"role": "system", "content": TOOL_COMPRESSION_SYSTEM_PROMPT},
            {"role": "user", "content": prompt},
        ]

        try:
            raw = await self._post_chat(conv)
            clean = extract_json_text(raw)
            validated = ToolSummaryOutput.model_validate_json(clean)
            return f"[Tool: {tool_name} | {validated.status}] {validated.summary}"
        except Exception as exc:
            logger.warning("Tool compression failed (%s); falling back to truncated output", exc)
            return f"[Tool: {tool_name}] {output_text[:max_chars_threshold]}... (truncated)"

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()


def create_memory_ai_adapter(settings: Settings | None = None) -> MemoryAIAdapter | None:
    """Create MemoryAIAdapter if MEMORY_AI_ENABLED is True, otherwise return None."""
    cfg = settings or get_settings()
    if not cfg.memory_ai_enabled:
        return None
    return MemoryAIAdapter(
        base_url=cfg.memory_ai_base_url,
        api_key=cfg.memory_ai_api_key,
        model=cfg.memory_ai_model,
        timeout=cfg.memory_ai_timeout,
        retry_once=cfg.memory_ai_retry_once,
    )

File: src/providers/memory-ai.ts
import type { MemoryAIOutput, ToolSummaryOutput } from "../models/memory";
import type { Env } from "../env";

const JSON_BLOCK_RE = /```(?:json)?\s*(?<json>\{.*\}|\[.*\])\s*```/s;

export const EXTRACTION_SYSTEM_PROMPT = `You are an AI Memory Compressor. Your role is ONLY to extract and compress memory items from conversation messages.
You NEVER answer the user directly.

Split multi-fact messages into ATOMIC facts. Each fact is ONE independent, self-contained statement (one entity, one attribute, one value). Do not merge multiple facts into a single bullet.

Output strictly valid JSON conforming to this schema:
{
  "summary": "Brief 1-2 sentence overview of what changed or happened",
  "facts": ["Stable facts about project/user/environment"],
  "decisions": ["Confirmed choices and decisions"],
  "constraints": ["Rules, restrictions, and negative requirements"],
  "preferences": ["User preferences and stylistic choices"],
  "goals": ["Active objectives and desired outcomes"],
  "architecture": ["System design and technical stack details"],
  "important_events": ["Key milestones, releases, or deployments"],
  "active_tasks": ["Pending or in-progress todos and tasks"],
  "obsolete_items": ["Previous facts/decisions that are now obsolete"],
  "contradictions": ["Previous items contradicted or superseded by new input"],
  "confidence": 0.95
}

Rules:
1. Extract only high-value information.
2. Each bullet must be a single atomic fact. For example, "Uses Cloudflare Workers, Hono, Turso, Upstash Redis, Groq" must become separate entries, not one.
3. When the user switches/changes a previously stated value (e.g. "we switched X to Y"), list the new value under the appropriate list and add the old one to "obsolete_items".
4. User statements and explicit decisions have highest authority.
5. Assistant speculation must NOT be recorded as confirmed facts/decisions.
6. Output ONLY valid JSON. No commentary outside the JSON.
`;

export const TOOL_COMPRESSION_SYSTEM_PROMPT = `You are a Tool Output Compressor. Your role is to summarize tool/command execution output.
Keep crucial exit statuses, error codes, file paths, IDs, and core outputs.
Remove repetitive logs, progress bars, and boilerplate.

Output strictly valid JSON:
{
  "summary": "Concise 1-3 sentence summary of the execution outcome",
  "key_points": ["Specific result item, path, or error"],
  "status": "success | error | warning"
}
`;

export function extractJsonText(text: string): string {
  const trimmed = text.trim();
  const match = trimmed.match(JSON_BLOCK_RE);
  if (match && match.groups && match.groups.json) {
    return match.groups.json.trim();
  }
  const firstBrace = trimmed.indexOf("{");
  const lastBrace = trimmed.lastIndexOf("}");
  if (firstBrace !== -1 && lastBrace !== -1 && lastBrace > firstBrace) {
    return trimmed.slice(firstBrace, lastBrace + 1);
  }
  return trimmed;
}

export class MemoryAIAdapter {
  private baseUrl: string;
  private apiKey?: string;
  private model: string;
  private retryOnce: boolean;

  constructor(options: {
    baseUrl: string;
    apiKey?: string;
    model: string;
    retryOnce?: boolean;
  }) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.model = options.model;
    this.retryOnce = options.retryOnce ?? true;
  }

  private authHeaders(): Record<string, string> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.apiKey) {
      headers["Authorization"] = `Bearer ${this.apiKey}`;
    }
    return headers;
  }

  private async postChat(messages: Array<{ role: string; content: string }>): Promise<string> {
    const url = `${this.baseUrl}/chat/completions`;
    const payload = {
      model: this.model,
      messages,
      temperature: 0.0,
    };
    const response = await fetch(url, {
      method: "POST",
      headers: this.authHeaders(),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Memory AI HTTP error ${response.status}: ${await response.text()}`);
    }

    const data = (await response.json()) as any;
    const choices = data?.choices || [];
    if (choices.length === 0) {
      throw new Error("No completion choices returned by Memory AI");
    }
    return String(choices[0]?.message?.content || "");
  }

  async extractMemory(
    messagesText: string,
    currentMemorySummary?: string | null
  ): Promise<MemoryAIOutput | null> {
    let userPrompt = `Delta Messages to analyze:\n${messagesText}`;
    if (currentMemorySummary) {
      userPrompt = `Current Memory State:\n${currentMemorySummary}\n\n${userPrompt}`;
    }

    const conversation: Array<{ role: string; content: string }> = [
      { role: "system", content: EXTRACTION_SYSTEM_PROMPT },
      { role: "user", content: userPrompt },
    ];

    let rawResponse: string | null = null;
    try {
      rawResponse = await this.postChat(conversation);
      return this.parseAndValidateMemory(rawResponse);
    } catch (parseErr) {
      if (!this.retryOnce) {
        return null;
      }
      try {
        const retryConv = [...conversation];
        if (rawResponse) {
          retryConv.push({ role: "assistant", content: rawResponse });
        }
        retryConv.push({
          role: "user",
          content: `Previous output was not valid JSON matching the schema: ${parseErr}. Please output ONLY the valid JSON object.`,
        });
        const retryRaw = await this.postChat(retryConv);
        return this.parseAndValidateMemory(retryRaw);
      } catch {
        return null;
      }
    }
  }

  private parseAndValidateMemory(rawText: string): MemoryAIOutput {
    const clean = extractJsonText(rawText);
    const parsed = JSON.parse(clean);
    return {
      summary: String(parsed.summary || ""),
      facts: Array.isArray(parsed.facts) ? parsed.facts.map(String) : [],
      decisions: Array.isArray(parsed.decisions) ? parsed.decisions.map(String) : [],
      constraints: Array.isArray(parsed.constraints) ? parsed.constraints.map(String) : [],
      preferences: Array.isArray(parsed.preferences) ? parsed.preferences.map(String) : [],
      goals: Array.isArray(parsed.goals) ? parsed.goals.map(String) : [],
      architecture: Array.isArray(parsed.architecture) ? parsed.architecture.map(String) : [],
      important_events: Array.isArray(parsed.important_events) ? parsed.important_events.map(String) : [],
      active_tasks: Array.isArray(parsed.active_tasks) ? parsed.active_tasks.map(String) : [],
      obsolete_items: Array.isArray(parsed.obsolete_items) ? parsed.obsolete_items.map(String) : [],
      contradictions: Array.isArray(parsed.contradictions) ? parsed.contradictions.map(String) : [],
      confidence: typeof parsed.confidence === "number" ? parsed.confidence : 0.95,
    };
  }

  async compressToolOutput(
    toolName: string,
    outputText: string,
    maxCharsThreshold = 300
  ): Promise<string> {
    if (outputText.length <= maxCharsThreshold) {
      return outputText;
    }

    const prompt = `Tool Name: ${toolName}\nRaw Output:\n${outputText}`;
    const conv = [
      { role: "system", content: TOOL_COMPRESSION_SYSTEM_PROMPT },
      { role: "user", content: prompt },
    ];

    try {
      const raw = await this.postChat(conv);
      const clean = extractJsonText(raw);
      const val = JSON.parse(clean) as ToolSummaryOutput;
      return `[Tool: ${toolName} | ${val.status || "success"}] ${val.summary || ""}`;
    } catch {
      return `[Tool: ${toolName}] ${outputText.slice(0, maxCharsThreshold)}... (truncated)`;
    }
  }
}

export function createMemoryAIAdapter(env: Env): MemoryAIAdapter | null {
  const isEnabled = String(env.MEMORY_AI_ENABLED).toLowerCase() === "true";
  if (!isEnabled) {
    return null;
  }
  const baseUrl = env.MEMORY_AI_BASE_URL || "https://openrouter.ai/api/v1";
  const model = env.MEMORY_AI_MODEL || "cheap-model";
  return new MemoryAIAdapter({
    baseUrl,
    apiKey: env.MEMORY_AI_API_KEY,
    model,
    retryOnce: true,
  });
}
